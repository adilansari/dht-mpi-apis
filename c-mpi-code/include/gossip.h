/*
 * (C) 2001 Clemson University and The University of Chicago
 *
 * See COPYING in top-level directory.
 */

/** \defgroup gossip gossip logging interface
 *
 * This is a basic application logging facility.  It uses printf style
 * formatting and provides several mechanisms for output.
 *
 * @{
 */

/** \file
 *
 *  Declarations for the gossip logging interface.
 */

#ifndef __GOSSIP_H
#define __GOSSIP_H

#ifndef _XOPEN_SOURCE
#define _XOPEN_SOURCE 500
#endif

#include <errno.h>
#include <stdio.h>
#include <stdarg.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <syslog.h>
#include <time.h>
#include <sys/time.h>
#ifdef HAVE_EXECINFO_H
#include <execinfo.h>
#endif
#ifdef __GOSSIP_ENABLE_MPI__
#include <mpi.h>
#endif

#ifndef __KERNEL__
#include <stdint.h>
#include <stdarg.h>
#include "syslog.h"
#endif
#include "cmpi-config.h"

// Masks...
#define MASK_ALL     ((uint64_t) -1)
#define MASK_NONE    ((uint64_t) 0)
#define MASK_DEBUG   ((uint64_t) 1 << 1)
#define MASK_MPIRPC  ((uint64_t) 1 << 2)
#define MASK_DHT     ((uint64_t) 1 << 3)
#define MASK_ADTS    ((uint64_t) 1 << 4)
#define MASK_DRIVER  ((uint64_t) 1 << 5)
#define MASK_CMPI    ((uint64_t) 1 << 6)
#define MASK_ARCH    ((uint64_t) 1 << 7)
#define MASK_STORAGE ((uint64_t) 1 << 8)
#define MASK_KDA     ((uint64_t) 1 << 9)
#define MASK_EXAMPLE ((uint64_t) 1 << 10)
#define MASK_CLIENT  ((uint64_t) 1 << 12)

/********************************************************************
 * Visible interface
 */

#define GOSSIP_BUF_SIZE 1024

/* what type of timestamp to place in msgs */
enum gossip_logstamp
{
    GOSSIP_LOGSTAMP_NONE = 0,
    GOSSIP_LOGSTAMP_USEC = 1,
    GOSSIP_LOGSTAMP_DATETIME = 2,
    GOSSIP_LOGSTAMP_THREAD = 3,
    GOSSIP_LOGSTAMP_MPI = 4,
};
#define GOSSIP_LOGSTAMP_DEFAULT GOSSIP_LOGSTAMP_MPI

#define gossip_debug_init() \
do {\
    char* __mask_str; \
    uint64_t __mask; \
    __mask_str = getenv("CMPI_DEBUG_MASK"); \
    __mask = (__mask_str ? gossip_set_debug_mask_human(__mask_str) : 0); \
    gossip_set_debug_mask(1, __mask); \
}while(0)

uint64_t gossip_set_debug_mask_human(
    const char *event_logging);

/* Keep a simplified version for the kmod */
#ifdef __KERNEL__

#ifdef GOSSIP_DISABLE_DEBUG
#define gossip_debug(mask, format, f...) do {} while(0)
#else
extern int gossip_debug_mask;

/* try to avoid function call overhead by checking masks in macro */
#define gossip_debug(mask, format, f...)                  \
do {                                                      \
    if (gossip_debug_mask & mask)                         \
    {                                                     \
        printk(format, ##f);                              \
    }                                                     \
} while(0)
#endif /* GOSSIP_DISABLE_DEBUG */

/* do file and line number printouts w/ the GNU preprocessor */
#define gossip_ldebug(mask, format, f...)                  \
do {                                                       \
    gossip_debug(mask, "%s: " format, __func__ , ##f); \
} while(0)

#define gossip_err printk
#define gossip_lerr(format, f...)                  \
do {                                               \
    gossip_err("%s line %d: " format, __FILE__ , __LINE__ , ##f); \
} while(0)

#else /* __KERNEL__ */

/* stdio is needed by gossip_debug_fp declaration for FILE* */
#include <stdio.h>

int gossip_enable_syslog(
    int priority);
int gossip_enable_stdout(
    void);
int gossip_enable_stderr(
    void);
int gossip_enable_file(
    const char *filename,
    const char *mode);
int gossip_disable(
    void);
int gossip_set_debug_mask(
    int debug_on,
    uint64_t mask);
int gossip_get_debug_mask(
    int *debug_on,
    uint64_t *mask);
int gossip_set_logstamp(
    enum gossip_logstamp ts);

void gossip_backtrace(void);

#ifdef __GNUC__

/* do printf style type checking if built with gcc */
int __gossip_debug(
    uint64_t mask,
    char prefix,
    const char *format,
    ...) __attribute__ ((format(printf, 3, 4)));
int gossip_err(
    const char *format,
    ...) __attribute__ ((format(printf, 1, 2)));
int __gossip_debug_va(
    uint64_t mask,
    char prefix,
    const char *format,
    va_list ap);
int gossip_debug_fp(
    FILE *fp,
    char prefix,
    enum gossip_logstamp ts,
    const char *format,
    ...) __attribute__ ((format(printf, 4, 5)));

#ifdef GOSSIP_DISABLE_DEBUG
#define gossip_debug(mask, format, f...) do {} while(0)
#define gossip_perf_log(format, f...) do {} while(0)
#define gossip_debug_enabled(__m) 0
#else
extern int gossip_debug_on;
extern int gossip_facility;
extern uint64_t gossip_debug_mask;

#define gossip_debug_enabled(__m) \
    (gossip_debug_on && (gossip_debug_mask & __m))

/* try to avoid function call overhead by checking masks in macro */
#define gossip_debug(mask, format, f...)                  \
do {                                                      \
    if ((gossip_debug_on) && (gossip_debug_mask & mask) &&\
        (gossip_facility))                                \
    {                                                     \
        __gossip_debug(mask, '?', format, ##f);           \
    }                                                     \
} while(0)
#define gossip_perf_log(format, f...)                     \
do {                                                      \
    if ((gossip_debug_on) &&                              \
        (gossip_debug_mask & GOSSIP_PERFCOUNTER_DEBUG) && \
        (gossip_facility))                                \
    {                                                     \
        __gossip_debug(GOSSIP_PERFCOUNTER_DEBUG, 'P',     \
            format, ##f);                                 \
    }                                                     \
} while(0)
#define gossip_do(mask, x)                                      \
    if ((gossip_debug_on) && (gossip_debug_mask & mask) &&      \
        (gossip_facility))                                      \
    { x; }

#endif /* GOSSIP_DISABLE_DEBUG */

/* do file and line number printouts w/ the GNU preprocessor */
#define gossip_ldebug(mask, format, f...)                  \
do {                                                       \
    gossip_debug(mask, "%s %s() line %d: " format, __FILE__, __func__ , __LINE__, ##f); \
} while(0)

#define gossip_lerr(format, f...)                  \
do {                                               \
    gossip_err("%s %s() line %d: " format, __FILE__ , __func__, __LINE__ , ##f); \
    gossip_backtrace();                            \
} while(0)
#else /* ! __GNUC__ */

int __gossip_debug(
    uint64_t mask,
    char prefix,
    const char *format,
    ...);
int __gossip_debug_stub(
    uint64_t mask,
    char prefix,
    const char *format,
    ...);
int gossip_err(
    const char *format,
    ...);

#ifdef GOSSIP_DISABLE_DEBUG
#define gossip_debug(__m, __f, f...) __gossip_debug_stub(__m, '?', __f, ##f);
#define gossip_ldebug(__m, __f, f...) __gossip_debug_stub(__m, '?', __f, ##f);
#define gossip_debug_enabled(__m) 0
#else
#define gossip_debug(__m, __f, f...) __gossip_debug(__m, '?', __f, ##f);
#define gossip_ldebug(__m, __f, f...) __gossip_debug(__m, '?', __f, ##f);
#define gossip_debug_enabled(__m) \
            ((gossip_debug_on != 0) && (__m & gossip_debug_mask))

#endif /* GOSSIP_DISABLE_DEBUG */

#define gossip_lerr gossip_err

#endif /* __GNUC__ */

#endif /* __KERNEL__ */

#endif /* __GOSSIP_H */

/* @} */

/*
 * Local variables:
 *  c-indent-level: 4
 *  c-basic-offset: 4
 * End:
 *
 * vim: ts=8 sts=4 sw=4 expandtab
 */
