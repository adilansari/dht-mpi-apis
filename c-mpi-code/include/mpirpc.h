
#ifndef MPIRPC_H
#define MPIRPC_H

#include <cmpi-config.h>

#include <mpi.h>

#include <stdbool.h>

#define MPIRPC_MAX_NAME 128
#define MPIRPC_MAX_ARGS 256
#define MPIRPC_PROCEED_NULL ((void (*)(MPIRPC*)) 1)

typedef struct
{
  MPI_Comm    comm;
  int         rank;
} MPIRPC_Node;

typedef struct
{
  MPI_Comm    comm;
  MPI_Request request;
} MPIRPC_Channel;

typedef struct mpirpc
{
  MPIRPC_Node target;
  int   status;
  char  name[MPIRPC_MAX_NAME];
  char  args[MPIRPC_MAX_ARGS];
  void* blob;
  int   blob_length;
  void* result;
  int   result_length;
  void* extras;
  void  (*proceed) (struct mpirpc* rpc);
  int   unique;
  bool  cancelled;

  // Internals:
  /**
     name_length is transferred via Isend().
  */
  int name_length;

  /**
     args_length is transferred via Isend().
  */
  int args_length;

  /**
     Array of Isend() requests to be freed on return.
  */
  MPI_Request request[8];
} MPIRPC;

/**
   Internal MPIRPC return value.  Used for garbage collection.
*/
typedef struct
{
  int   unique;
  int   length;
  void* result;
  MPI_Request request[4];
} MPIRPC_Value;

/**
   Defines MPIRPC handler "f"
   Parameters:
      the target node
      the uniqifier
      the args
      the blob
      the blob length
 */
#define MPIRPC_Handler_pointer(f) void \
        (*f)(MPIRPC_Node,int,char*,void*,int)

#define MPIRPC_Handler(f) void \
        f(MPIRPC_Node caller, int unique, char* args, void* blob, int blob_length)

//// API...

void MPIRPC_Init(void);

bool MPIRPC_Check(void);

void MPIRPC_Recv(MPI_Comm comm);

void MPIRPC_Register(char* name, MPIRPC_Handler_pointer(f));

bool MPIRPC_Comm_add(MPI_Comm comm);

void MPIRPC_Close(MPIRPC_Channel* channel);

MPIRPC* MPIRPC_Call(MPIRPC_Node target, char* name, char* args,
                    void* extras, void (*proceed)(MPIRPC*));

MPIRPC* MPIRPC_Call_blob(MPIRPC_Node target, char* name, char* args,
                         void* blob, int blob_length,
                         void* extras, void (*proceed)(MPIRPC*));

void* MPIRPC_Block(MPIRPC_Node target, char* name, char* args);

void* MPIRPC_Block_blob(MPIRPC_Node target, char* name, char* args,
                        void* blob, int blob_length);

void* MPIRPC_Wait(MPIRPC* rpc);

void MPIRPC_Null(MPIRPC_Node caller, int unique);

void MPIRPC_Return(MPIRPC_Node caller, int unique,
                   void* result, int rlength);

MPIRPC_Node* MPIRPC_Node_create(MPI_Comm comm, int rank);

MPIRPC_Node* MPIRPC_Node_make(MPI_Comm comm, int rank,
                              MPIRPC_Node* node);

void MPIRPC_Garbage_collect(void);
bool MPIRPC_Garbage_collect_value(void);

// struct list* MPIRPC_Comms(void);

void MPIRPC_Dump(MPIRPC* rpc);

void MPIRPC_Dump_portmap(void);

void MPIRPC_Dump_channels(void);

void MPIRPC_Flush_returns(void);

char* MPIRPC_Comm_get_name(MPI_Comm comm);

void MPIRPC_Free(MPIRPC* rpc);

void MPIRPC_Destroy(MPIRPC* rpc);

void MPIRPC_Snooze(void);

void MPIRPC_Snooze_reset(void);

void MPIRPC_Snooze_set(int s);

//// Data structure helpers...

int MPIRPC_Node_cmp(void* node1, void* node2);

int MPIRPC_Channel_cmp(void* c1, void* c2);

int MPIRPC_Channel_comm_cmp(void* c1, void* c2);

MPIRPC_Channel* MPIRPC_Channel_for_node(MPIRPC_Node node);

void MPIRPC_Finalize(void);

/**
   MPIRPC stati.
*/
enum
{
  MPIRPC_STATUS_PROTO,
  MPIRPC_STATUS_CALLED,
  MPIRPC_STATUS_RETURNED
};

/**
   Tags used by MPIRPC.
*/
enum
{
  MPIRPC_TAG_CONTROL,
  MPIRPC_TAG_UNIQUE,
  MPIRPC_TAG_NAME,
  MPIRPC_TAG_LENGTH,
  MPIRPC_TAG_ARGS,
  MPIRPC_TAG_RESULT
};

enum
{
  MPIRPC_MSG_CALL       = 12,
  MPIRPC_MSG_RETURN     = 13,
  MPIRPC_MSG_DISCONNECT = 14,
  MPIRPC_MSG_OK         = 15,
  MPIRPC_MSG_FAIL       = 16
};

#endif
