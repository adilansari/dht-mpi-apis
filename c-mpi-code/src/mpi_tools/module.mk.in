
DIR := src/mpi_tools

MPITOOLS_SRC :=
MPITOOLS_SRC += $(DIR)/mpi_tools.c
MPITOOLS_SRC += $(DIR)/io_tools.c

MPITOOLS_OBJS = $(patsubst %.c, %.o, $(MPITOOLS_SRC))
