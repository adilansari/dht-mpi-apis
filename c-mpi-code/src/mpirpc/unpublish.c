
/**
   Simply unpublish a name from the MPI-2 namespace. 
*/

#include <stdio.h> 
#include <stdlib.h>

#include <mpi.h>

void
printhelp(void)
{
  printf("usage: unpublish <name> \n");
  printf("returns 0 on success, 1 on error \n"); 
}

void
crash(int value)
{
  MPI_Finalize(); 
  exit(value); 
}

int
main(int argc, char* argv[])
{
  MPI_Init(&argc, &argv);

  if (argc != 2 ||
      argv[1][0] == '-')
  {
    printhelp(); 
    crash(1); 
  }

  char* name = argv[1];
  char  port[MPI_MAX_PORT_NAME];
  int   err = MPI_Lookup_name(name, MPI_INFO_NULL, port);
  if (err == MPI_SUCCESS)
  {
    printf("%s found.  Unpublishing... \n", name); 
    MPI_Unpublish_name(name, MPI_INFO_NULL, port); 
  }
  else
  {
    printf("%s not found! \n", name);
    crash(1); 
  }

  MPI_Finalize();

  return 0; 
}
