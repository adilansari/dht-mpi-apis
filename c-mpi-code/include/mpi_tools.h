
#ifndef MPI_TOOLS
#define MPI_TOOLS

#include <assert.h>
#include <math.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <sys/stat.h>
#include <sys/time.h>

#include <mpi.h>

#ifdef ENABLE_DEBUG
#ifdef HAVE_EXEC_INFO_H
#include <execinfo.h>
#endif
#endif

#include <gossip.h>

extern int mpi_rank;
extern int mpi_size;

extern int debug_rank;

extern FILE* debug_file;

/**
   Quick setup
 */
void whoami(void);

char* heap(char* s);
char* iheap(int i);
char* xheap(int i);

/**
   Execute code line or block in rank order.
*/
#define sequential(code)                                                \
  {                                                                     \
    int sequential_i;                                                   \
    for (sequential_i = 0; sequential_i < mpi_size;                     \
         sequential_i++)                                                \
    {                                                                   \
      if (mpi_rank == sequential_i)                                     \
      {                                                                 \
        code;                                                           \
      }                                                                 \
      MPI_Barrier(MPI_COMM_WORLD);                                      \
    }                                                                   \
  }

void printbinary(unsigned int number);
void printdata(char* tag, char* data, int length);
void fprintdata(FILE* file, char* tag, char* data, int length);
int  flip2(int i);
unsigned int ilog2(unsigned int j);
unsigned int isqrt(unsigned int j);

/**
   Return a random number in [0,n).
*/
int rand_lt(int n);

/**
   Return a random double in [0,n).
*/
double random_double(int n);

/**
   Concatenate char* s at char* p, update p to the end.
*/
#define cat(p, s) p += sprintf(p, "%s", s);

#ifdef ENABLE_DEBUG
// #warning "ENABLE_DEBUG"
#define DEBUG(x) x
#else
#define DEBUG(x)
#endif

#ifdef ENABLE_DEBUG
/**
   Dump the stack and exit.
*/
void
dump_stack(void);
#endif

#define OK NOTE("ok")

void setup_debug_file(void);

/**
   Simplify debugging statements.
 */
void note(char* msg);
void note_b(char* msg, int b);
void note_d(char* msg, double d);
void note_i(char* msg, int i);
void note_ii(char* msg, int i, int j);
void note_s(char* msg, char* s);
void note_ss(char* msg, char* s1, char* s2);
void note_p(char* msg, void* p);
void note_u(char* msg, unsigned int u);
void note_x(char* msg, int i);
void note_xi(char* msg, int i, int j);
void note_sx(char* msg, char* s, int i);
void note_sxi(char* msg, char* s, int x, int i);
// void note_ssxi(char* msg, char* s, int x, int i);

void show_b(char* tag, int b);
void show_d(char* tag, double d);
void show_i(char* tag, int i);
void show_ii(char* tag_i, int i, char* tag_j, int j);
void show_l(char* tag, long l);
void show_s(char* tag, char* s);
void show_u(char* tag, unsigned int u);
void show_x(char* tag, int x);
void show_p(char* tag, void* p);

void  timestamp(char* label, char* msg);
char* timestring(char* label, char* msg);

#define NOTE(m)                DEBUG(note(m))
#define NOTE_B(m,b)            DEBUG(note_b(m,b))
#define NOTE_D(m,d)            DEBUG(note_d(m,d))
#define NOTE_I(m,i)            DEBUG(note_i(m,i))
#define NOTE_II(m,i,j)         DEBUG(note_ii(m,i,j))
#define NOTE_U(m,u)            DEBUG(note_u(m,u))
#define NOTE_P(m,s)            DEBUG(note_p(m,))
#define NOTE_S(m,s)            DEBUG(note_s(m,s))
#define NOTE_SX(m,s,i)         DEBUG(note_sx(m,s,i))
#define NOTE_SXI(m,s,x,i)      DEBUG(note_sxi(m,s,x,i))
// #define NOTE_SSXI(m,s1,s2,x,i) DEBUG(note_ssxi(m,s1,s2,x,i))
#define NOTE_X(m,i)            DEBUG(note_x(m,i))
#define NOTE_XI(m,i,j)         DEBUG(note_xi(m,i,j))

#define NOTE_F DEBUG(fprintf(debug_file, "[%i] %s()... \n",     \
                             debug_rank, __func__);             \
                     fflush(debug_file);)
#define DONE DEBUG(fprintf(debug_file, "[%i] %s() done. \n",     \
                          debug_rank, __func__);        \
                   fflush(debug_file);)
#define NOTE_FI(i)                              \
  DEBUG(fprintf(debug_file, "[%i] %s(%i)... \n",             \
               debug_rank, __func__, i);        \
        fflush(stdout);)
#define NOTE_FX(i) DEBUG(fprintf(debug_file, "[%i] %s(%X)... \n",            \
                                debug_rank, __func__, i);       \
                         fflush(stdout);)
#define NOTE_FIS(i,s) DEBUG(fprintf(debug_file, "[%i] %s(%i,%s)... \n",      \
                                   debug_rank, __func__, i, s); \
                            fflush(stdout);)
#define NOTE_FXIS(x,i,s)                        \
  DEBUG(fprintf(debug_file, "[%i] %s(%X,%i,%s)... \n",       \
               debug_rank, __func__, x, i, s);  \
        fflush(stdout);)
#define NOTE_FXI(i,j)                           \
  DEBUG(fprintf(debug_file, "[%i] %s(%X,%i)... \n",          \
               debug_rank, __func__, i, j);     \
        fflush(stdout);)
#define NOTE_FXS(i,s)                           \
  DEBUG(fprintf(debug_file, "[%i] %s(%X,%s)... \n",          \
               debug_rank, __func__, i, s);     \
        fflush(stdout););
#define NOTE_FXX(i,j) DEBUG(fprintf(debug_file, "[%i] %s(%X,%X)... \n",      \
                                   debug_rank, __func__, i, j); \
                            fflush(stdout);)
#define NOTE_FS(s) DEBUG(fprintf(debug_file, "[%i] %s(%s)... \n",            \
                                debug_rank, __func__, s);       \
                         fflush(stdout);)
#define NOTE_FSS(s1,s2)                         \
  DEBUG(fprintf(debug_file, "[%i] %s(%s,%s)... \n",          \
               debug_rank, __func__, s1, s2);   \
        fflush(stdout);)
#define NOTE_FSSS(s1,s2,s3)                     \
  DEBUG(fprintf(debug_file, "[%i] %s(%s,%s,%s)... \n",       \
               debug_rank, __func__,            \
               s1, s2, s3);                     \
        fflush(stdout);)
#define NOTE_FSSSI(s1,s2,s3,i)                          \
  DEBUG(fprintf(debug_file, "[%i] %s(%s,%s,%s,%i)... \n",            \
               debug_rank, __func__, s1, s2, s3, i);    \
        fflush(stdout);)

///// Simple variable debugging...
#define SHOW_B(b)    DEBUG(show_b(#b, b))
#define SHOW_D(d)    DEBUG(show_d(#d, d))
#define show_i(i)    note_i(#i ": ", i)
#define SHOW_I(i)    DEBUG(show_i(i))
#define SHOW_II(i,j) DEBUG(show_ii(#i, i, #j, j))
#define SHOW_L(l)    DEBUG(show_l(#l, l))
#define SHOW_U(u)    DEBUG(show_u(#u, u))
#define show_s(s)    note_s(#s ": ", s)
#define SHOW_S(s)    DEBUG(show_s(s))
#define SHOW_X(x)    DEBUG(show_x(#x, x))
#define SHOW_P(p)    DEBUG(show_p(#p, p))
#define SHOW         SHOW_S

//// Functions...
#define show_fi(i)                              \
       fprintf(debug_file, "[%i] %s(%s=%i)... \n",          \
               debug_rank, __func__, #i, i);    \
        fflush(stdout);
#define SHOW_FI(i) DEBUG(show_fi(i))
#define SHOW_FIS(i,s)                                   \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%i,%s=%s)... \n",            \
               debug_rank, __func__, #i, i, #s, s);     \
        fflush(stdout);)
#define SHOW_FISS(i,s1,s2)                                      \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%i,%s=%s,%s=%s)... \n",              \
               debug_rank, __func__, #i, i, #s1, s1, #s2, s2);  \
        fflush(stdout);)
#define SHOW_FISX(i,s,x)                                                \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%i,%s=%s,%s=%X)... \n",         \
                debug_rank, __func__, #i, i, #s, s, #x, x);             \
        fflush(stdout);)

#define show_fs(s) fprintf(debug_file,           \
                           "[%i] %s(%s=%s)... \n",          \
               debug_rank, __func__, #s, s);    \
        fflush(stdout);
#define SHOW_FS(s)                              \
  DEBUG(show_fs(s))

#define SHOW_FX(x)                                \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%X)... \n",            \
               debug_rank, __func__, #x, x);      \
        fflush(stdout);)
#define SHOW_FII(i1,i2)                                     \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%i,%s=%i)... \n",                \
               debug_rank, __func__, #i1, i1, #i2, i2);     \
        fflush(stdout);)

#define show_fiii(i1,i2,i3)                              \
      fprintf(debug_file, "[%i] %s(%s=%i,%s=%i,%s=%i)... \n",  \
              debug_rank, __func__, #i1, i1, #i2, i2, #i3, i3);   \
        fflush(stdout);
#define SHOW_FIII(i1,i2,i3) DEBUG(show_fiii(i1,i2,i3))

#define show_fsi(s,i)  \
      fprintf(debug_file, "[%i] %s(%s=%s,%s=%i)... \n",  \
                debug_rank, __func__, #s, s, #i, i);     \
        fflush(stdout);
#define SHOW_FSI(s,i) DEBUG(show_fsi(s,i))

#define show_fsii(s,i1,i2)                                     \
  fprintf(debug_file, "[%i] %s(%s=%s,%s=%i,%s=%i)... \n",      \
              debug_rank, __func__, #s, s, #i1, i1, #i2, i2);  \
  fflush(stdout);
#define SHOW_FSII(s,i1,i2) DEBUG(show_fsii(s,i1,i2))

#define SHOW_FXI(x,i)                                   \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%X,%s=%i)... \n",            \
               debug_rank, __func__, #x, x, #i, i);     \
        fflush(stdout);)

#define SHOW_FXIS(x,i,s)                                \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%X,%s=%i,%s=%s)... \n",      \
               debug_rank, __func__,                    \
               #x, x, #i, i, #s, s);                    \
        fflush(stdout);)

#define SHOW_FXIIS(x,i1,i2,s)                                \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%X,%s=%i,%s=%i,%s=%s)... \n",      \
               debug_rank, __func__,                    \
               #x, x, #i1, i1, #i2, i2, #s, s);                    \
        fflush(stdout);)

#define SHOW_FSS(s1,s2)                                 \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%s,%s=%s)... \n",            \
               debug_rank, __func__, #s1, s1, #s2, s2); \
        fflush(stdout);)
#define SHOW_FSSI(s1,s2,i)                                             \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%s,%s=%s,%s=%i)... \n",        \
                debug_rank, __func__, #s1, s1, #s2, s2, #i, i);        \
        fflush(stdout);)
#define SHOW_FISSI(i1,s1,s2,i2)                                \
  DEBUG(fprintf(debug_file,                                        \
                "[%i] %s(%s=%i,%s=%s,%s=%s,%s=%i)... \n",    \
                debug_rank, __func__, #i1, i1, #s1, s1,   \
                #s2, s2, #i2, i2); \
        fflush(stdout);)
#define SHOW_FSISSI(s1,i1,s2,s3,i2)                                \
  DEBUG(fprintf(debug_file,                                        \
                "[%i] %s(%s=%s,%s=%i,%s=%s,%s=%s,%s=%i)... \n",    \
                debug_rank, __func__, #s1, s1, #i1, i1, #s2, s2,   \
                #s3, s3, #i2, i2); \
        fflush(stdout);)
#define SHOW_FP(p)                              \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%p)... \n",          \
               debug_rank, __func__, #p, p);    \
        fflush(stdout);)

//// Functions with return values...

#define SHOW_FS_X(s,x)                                                  \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%s) -> %X... \n",               \
                debug_rank, __func__, #s, s, x);                        \
        fflush(stdout);)
#define SHOW_FIFS_S(i,f,s1,s2)                                       \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%i,%s(%s) -> %s)... \n",     \
                debug_rank, __func__, #i, i, f, s1, s2);             \
        fflush(stdout);)
#define SHOW_FIS_X(i,s,x)                                             \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%i,%s=%s) -> %X... \n",       \
                debug_rank, __func__, #i, i, #s, s, x);               \
        fflush(stdout);)
#define SHOW_FXI_B(x,i,b)                                               \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%X,%s=%i) -> %s... \n",         \
                debug_rank, __func__,                                   \
                #x, x, #i, i, b ? "true" : "false");                    \
        fflush(stdout);)


//// Functions of functions...

#define SHOW_FIF(i, f)                               \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%i,%s())... \n",           \
               debug_rank, __func__, #i, i, f);       \
        fflush(stdout);)
#define SHOW_FIFS(i, f, s)                             \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%i,%s(%s))... \n",           \
               debug_rank, __func__, #i, i, f, s);      \
        fflush(stdout);)

#define TIMESTAMP(m) note_i(m, time());

#define DUMPSTACK DEBUG(dump_stack());

char* file_string(char* file);

//// Memory usage debugging:
#ifdef DMALLOC
#include <dmalloc.h>
#define DMALLOC_SETUP(x) x
void dmalloc_setup(void);
#else
#define DMALLOC_SETUP(x)
#define dmalloc_setup(x)
#endif

#define DMALLOC_COMPLETE                              \
  DMALLOC_SETUP(                                      \
    {                                                 \
      SHOW_S(dmalloc_logpath);                        \
      dmalloc_shutdown();                             \
      dmalloc_log_unfreed();                          \
    });

/**
   Return a number in [0,n-1].
*/
int roll(int n);

#endif
