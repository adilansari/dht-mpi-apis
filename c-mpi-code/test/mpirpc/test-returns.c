
/**
   Simple two-processor argument test
   for non-0-length args and
   for NULL, 0-length, and non-0-length return values.
*/

#include <mpi_tools.h>
#include <gossip.h>

#include <mpirpc.h>

#include "test_helpers.h"

void
handle_test(MPIRPC_Node caller, int unique, char* args,
            void* blob, int blob_length)
{
  int a;
  sscanf(args, "%i", &a);
  show_fsi(args, a);
  char* result = malloc(4);
  SHOW_P(result);

  if (a == 0)
  {
    MPIRPC_Null(caller, unique);
  }
  else if (a == 1)
  {
    MPIRPC_Return(caller, unique, result, 0);
  }
  else if (a == 2)
  {
    strcpy(result, "ok\n");
    MPIRPC_Return(caller, unique, result, 4);
  }
}

int
main(int argc, char* argv[])
{
  gossip_set_debug_mask(1, MASK_MPIRPC);

  MPI_Init(&argc, &argv);
  MPIRPC_Init();
  whoami();
  setup_debug_file();

  MPIRPC_Comm_add(MPI_COMM_WORLD);

  MPIRPC_Node neighbor;
  if (mpi_rank == 0)
  {
    char* result;
    MPIRPC_Node_make(MPI_COMM_WORLD, 1, &neighbor);
    result = MPIRPC_Block(neighbor, "test", "0");
    show_s(result);
    if (result)
      free(result);
    result = MPIRPC_Block(neighbor, "test", "1");
    show_s(result);
    if (result)
      free(result);
    result = MPIRPC_Block(neighbor, "test", "2");
    show_s(result);
    if (result)
      free(result);
  }
  else
  {
    MPIRPC_Node_make(MPI_COMM_WORLD, 0, &neighbor);
    MPIRPC_Register("test", handle_test);
    while (! MPIRPC_Check());
    while (! MPIRPC_Check());
    while (! MPIRPC_Check());
  }

  printf("Normal exit.\n");
  MPIRPC_Finalize();
  MPI_Finalize();

  return 0;
}
