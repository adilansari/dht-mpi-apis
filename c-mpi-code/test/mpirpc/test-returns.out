Debug to stdout
[0] MPIRPC_Block(test)... 
[0] MPIRPC_Block_blob(test)... 
[0] MPIRPC_Call_blob(target.rank=1,name=test,args=0,rpc->unique=1)... 
[1] MPIRPC_Request(name=test,caller.rank=0)... 
[1] handle_test(args=0,a=0)... 
[1] result: 0x1925da0
[1] MPIRPC_Return(caller.rank=0,unique=1,length=-1)... 
[0] MPIRPC_Retrieve(sender.rank=1,unique=1)... 
[0] Unblocked: test
[0] MPIRPC_Free(rpc->name=test,rpc->unique=1)... 
[0] result: (null)
[0] MPIRPC_Block(test)... 
[0] MPIRPC_Block_blob(test)... 
[0] MPIRPC_Call_blob(target.rank=1,name=test,args=1,rpc->unique=2)... 
[1] MPIRPC_Request(name=test,caller.rank=0)... 
[1] handle_test(args=1,a=1)... 
[1] result: 0x1925df0
[1] MPIRPC_Return(caller.rank=0,unique=2,length=0)... 
[0] MPIRPC_Retrieve(sender.rank=1,unique=2)... 
[0] Unblocked: test
[0] MPIRPC_Free(rpc->name=test,rpc->unique=2)... 
[0] result: (null)
[0] MPIRPC_Block(test)... 
[0] MPIRPC_Block_blob(test)... 
[0] MPIRPC_Call_blob(target.rank=1,name=test,args=2,rpc->unique=3)... 
[1] MPIRPC_Request(name=test,caller.rank=0)... 
[1] handle_test(args=2,a=2)... 
[1] result: 0x1925df0
[1] MPIRPC_Return(caller.rank=0,unique=3,length=4)... 
Normal exit.
[1] MPIRPC_Finalize()... 
[0] MPIRPC_Retrieve(sender.rank=1,unique=3)... 
[0] Unblocked: test
[0] MPIRPC_Free(rpc->name=test,rpc->unique=3)... 
[0] result: ok

Normal exit.
[0] MPIRPC_Finalize()... 
